[Style] Convert list-style-type to use strong style types

weinig opened the pull request:
#### 2d72924c6decf6fffba344a7fe8aca39d411cb06
<pre>
[Style] Convert list-style-type to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=296073">https://bugs.webkit.org/show_bug.cgi?id=296073</a>

Reviewed by Tim Nguyen.

Updates the existing ListStyleType type to be a strong style type, adding
variant-like conformance to remove need for custom extractor and serialization.
Additional IPC specific construction was needed for WebKit&apos;s IPC coders.

ListStyleType was incorrectly being used by the counters code where a
CounterStyle (a subset of ListStyleType) was actually needed, so this also
adds a CounterStyle type and adopts it. This allowed the removal of a bunch
of assertions that are now impossible.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilityList.cpp:
* Source/WebCore/css/CSSCounterStyle.cpp:
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSCounterValue.cpp:
* Source/WebCore/css/CSSCounterValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Content.cpp:
* Source/WebCore/editing/FontAttributes.h:
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
* Source/WebCore/rendering/RenderCounter.cpp:
* Source/WebCore/rendering/RenderCounter.h:
* Source/WebCore/rendering/RenderListMarker.cpp:
* Source/WebCore/rendering/style/CounterContent.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/counter-styles/StyleCounterStyle.cpp: Renamed from Source/WebCore/rendering/style/ListStyleType.h.
* Source/WebCore/style/values/counter-styles/StyleCounterStyle.h: Renamed from Source/WebCore/rendering/style/ListStyleType.cpp.
* Source/WebCore/style/values/lists/StyleListStyleType.cpp: Added.
* Source/WebCore/style/values/lists/StyleListStyleType.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/297551@main">https://commits.webkit.org/297551@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/feeff6f860766335988917cd2c1807bc05405a50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112096 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31769 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118123 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62351 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32451 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85167 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35834 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115043 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25920 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65595 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19019 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61981 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95308 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19081 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121445 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39134 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29135 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93999 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39515 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97110 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93822 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23956 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39030 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16820 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35172 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39027 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44546 "Found 1 new failure in rendering/svg/SVGTextLayoutEngineSpacing.h") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38662 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41993 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40378 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
```diff
@@ -39,37 +39,45 @@
-    return ordered ? NSTextListMarkerDecimal : NSTextListMarkerDisc;
+    return WTF::switchOn(styleType,
+        [&](const Style::CounterStyle& counterStyle) {
+            if (counterStyle == CSSValueDisc)
+                return NSTextListMarkerDisc;
+            if (counterStyle == CSSValueCircle)
+                return NSTextListMarkerCircle;
+            if (counterStyle == CSSValueSquare)
+                return NSTextListMarkerSquare;
+            if (counterStyle == CSSValueDecimal)
+                return NSTextListMarkerDecimal;
+            if (counterStyle == CSSValueOctal)
+                return NSTextListMarkerOctal;
+            if (counterStyle == CSSValueLowerRoman)
+                return NSTextListMarkerLowercaseRoman;
+            if (counterStyle == CSSValueUpperRoman)
+                return NSTextListMarkerUppercaseRoman;
+            if (counterStyle == CSSValueLowerAlpha)
+                return NSTextListMarkerLowercaseAlpha;
+            if (counterStyle == CSSValueUpperAlpha)
+                return NSTextListMarkerUppercaseAlpha;
+            if (counterStyle == CSSValueLowerLatin)
+                return NSTextListMarkerLowercaseLatin;
+            if (counterStyle == CSSValueUpperLatin)
+                return NSTextListMarkerUppercaseLatin;
+            if (counterStyle == CSSValueLowerHexadecimal)
+                return NSTextListMarkerLowercaseHexadecimal;
+            if (counterStyle == CSSValueUpperHexadecimal)
+                return NSTextListMarkerUppercaseHexadecimal;
+
+            // The remaining web-exposed list style types have no Cocoa equivalents.
+            // Fall back to default styles for ordered and unordered lists.
+            return ordered ? NSTextListMarkerDecimal : NSTextListMarkerDisc;
```
Could this be a switch statement? or does the conversion from Style::CounterStyle& to CSSValueID fail?

(Also we should really stop generating CSSValueID for these and just compare the names with strings, but probably a separate change)

weinig replied:
Unfortunately can't be a switch statement, CounterStyle stores an AtomString.

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -324,36 +324,33 @@ void RenderListMarker::updateContent()
+                .textDirection = TextDirection::LTR,
+            };
+        },
+        [&](const AtomString& identifier) {
```
nit: weird indent
```suggestion
    WTF::switchOn(style().listStyleType(),
        [&](const CSS::Keyword::None&) {
            m_textContent = {
                .textWithSuffix = " "_s,
                .textWithoutSuffixLength = 0,
                .textDirection = TextDirection::LTR,
            };
        },
        [&](const AtomString& identifier) {
```

weinig replied:
Oops.

nt1m commented on Source/WebCore/style/values/lists/StyleListStyleType.cpp:
```diff
@@ -0,0 +1,119 @@
+    if (!primitiveValue)
+        return CSS::Keyword::None { };
+
+    if (primitiveValue->isValueID()) {
+        if (primitiveValue->valueID() == CSSValueNone)
+            return CSS::Keyword::None { };
+        return CounterStyle { { AtomString { primitiveValue->stringValue() } } };
+    }
+    if (primitiveValue->isCustomIdent())
+        return CounterStyle { { AtomString { primitiveValue->stringValue() } } };
```
This logic could be a more simple if we stopped generating CSSValueID for the built-in ones (but probably a separate change here too, since it was pre-existing logic)

weinig committed:
[Style] Convert list-style-type to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=296073

Reviewed by Tim Nguyen.

Updates the existing ListStyleType type to be a strong style type, adding
variant-like conformance to remove need for custom extractor and serialization.
Additional IPC specific construction was needed for WebKit's IPC coders.

ListStyleType was incorrectly being used by the counters code where a
CounterStyle (a subset of ListStyleType) was actually needed, so this also
adds a CounterStyle type and adopts it. This allowed the removal of a bunch
of assertions that are now impossible.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilityList.cpp:
* Source/WebCore/css/CSSCounterStyle.cpp:
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSCounterValue.cpp:
* Source/WebCore/css/CSSCounterValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Content.cpp:
* Source/WebCore/editing/FontAttributes.h:
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
* Source/WebCore/rendering/RenderCounter.cpp:
* Source/WebCore/rendering/RenderCounter.h:
* Source/WebCore/rendering/RenderListMarker.cpp:
* Source/WebCore/rendering/style/CounterContent.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/counter-styles/StyleCounterStyle.cpp: Renamed from Source/WebCore/rendering/style/ListStyleType.h.
* Source/WebCore/style/values/counter-styles/StyleCounterStyle.h: Renamed from Source/WebCore/rendering/style/ListStyleType.cpp.
* Source/WebCore/style/values/lists/StyleListStyleType.cpp: Added.
* Source/WebCore/style/values/lists/StyleListStyleType.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/297551@main
